src/openrct2/world/Location.hpp

```diff
@@ -158,15 +158,173 @@ struct TileCoordsXYZ
  * 3 is Y-decreasing
  * Direction is not used to model up/down, or diagonal directions.
  */
-typedef uint8_t Direction;
+struct Direction
+{
+    enum : uint8_t
+    {
+        WEST = 0,
+        NORTH,
+        EAST,
+        SOUTH,
+        INVALID = 0xFF, // Used to mark invalid direction.
+    } direction;
+
+    Direction() = default;
+
+    constexpr Direction(decltype(direction) dir)
+        : direction(dir)
+    {
+    }
+
+    constexpr Direction(const uint8_t dir)
+        : direction(static_cast<decltype(direction)>(dir & 3))
+    {
+    }
+
+    bool operator==(const Direction& other) const
+    {
+        return other.direction == direction;
+    }
+
+    bool operator!=(const Direction& other) const
+    {
+        return other.direction != direction;
+    }
+
+    bool operator==(decltype(direction) dir) const
+    {
+        return dir == direction;
+    }
+
+    bool operator!=(decltype(direction) dir) const
+    {
+        return dir != direction;
+    }
+
+    Direction& operator=(const decltype(direction) dir)
+    {
+        direction = dir;
+        return *this;
+    }
+
+    Direction operator-(const Direction& dir)
+    {
+        Direction res;
+        res.direction = static_cast<decltype(direction)>(direction - dir.direction);
+        res.Normalise();
+        return res;
+    }
+
+    Direction operator-(const decltype(direction) dir)
+    {
+        Direction res;
+        res.direction = static_cast<decltype(direction)>(direction - dir);
+        res.Normalise();
+        return res;
+    }
+
+    Direction operator+(const Direction& dir)
+    {
+        Direction res;
+        res.direction = static_cast<decltype(direction)>(direction + dir.direction);
+        res.Normalise();
+        return res;
+    }
+
+    Direction operator+(const decltype(direction) dir)
+    {
+        Direction res;
+        res.direction = static_cast<decltype(direction)>(direction + dir);
+        res.Normalise();
+        return res;
+    }
+
+    Direction& operator++()
+    {
+        return RotateForward();
+    }
+
+    Direction& operator--()
+    {
+        return RotateBackward();
+    }
+
+    Direction& Reverse()
+    {
+        direction = static_cast<decltype(direction)>(direction ^ 2);
+        return *this;
+    }
+
+    Direction& Mirror()
+    {
+        direction = static_cast<decltype(direction)>(0 - direction);
+        Normalise();
+        return *this;
+    }
+
+    Direction& Rotate(int32_t times)
+    {
+        direction = static_cast<decltype(direction)>(direction * times);
+        Normalise();
+        return *this;
+    }
+
+    Direction& Normalise()
+    {
+        direction = static_cast<decltype(direction)>(direction & 3);
+        return *this;
+    }
```

Comment:
I don't think this should mutate, rather it should return a new one. Also why is this necessary? Surely it should be impossible to construct a Direction with an invalid number?

Reply:
Somewhat early to review the code, a lot of things are currently broken and I need to make sure that everything works first before I get to the cleanup.